### ISSUE TITLE
Different speed between BatchNorm1d and BatchNorm2d

### ISSUE THREAD

**Issue by prclibo** [pytorch/pytorch#38915]
Hi I assume BatchNorm1d and BatchNorm2d (or 3d) should have same performance. However when normalizing a tensor of shape `(L, C) = (B*H*W, C)` by BatchNorm1d, the speed is >5x slower than BatchNorm2d on a shape of `(B, C, H, W)`. Is this expected or is there a way to speed up?

Reproducer:

```
import torch

def time_bn(bn, x):
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    x = bn(x)
    end.record()
    torch.cuda.synchronize()
    print(start.elapsed_time(end))
    return x

shape = (50, 50, 50, 50)
a = torch.rand(shape)
b = a.flatten(0, -2)

bna = torch.nn.BatchNorm2d(50)
bnb = torch.nn.BatchNorm1d(50)

time_bn(bna, a)
time_bn(bnb, b)
```

Output:
```
11.409407615661621
89.55596923828125
```

## Environment

```
 - PyTorch Version: tested on 1.4/1.5
 - OS: Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.7/3.8
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: GTX 1070
 - Any other relevant information:
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel @csarofeen @ptrblck @xwang233 @albanD @mruberry @VitalyFedyunin

**Comment by ezyang** [comment]
Can you run the program under `nvprof` and show us which kernels are being run in each case?

**Comment by prclibo** [comment]
@ezyang Note that the above code is on CPU. For GPU version, such speed difference also similarly exists as follows. `nvprof` shows different kernels used.

Code:
```python
import torch

def time_bn(bn, x):
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    x = bn(x)
    end.record()
    torch.cuda.synchronize()
    # print('++++++++++ CUDA Time +++++++++++')
    # print(start.elapsed_time(end))
    return x

shape = (50, 50, 50, 50)
a = torch.rand(shape).cuda()
b = a.flatten(0, -2) + 1

# Commented resp during nvprof.
bna = torch.nn.BatchNorm2d(50).cuda()
time_bn(bna, a)
bnb = torch.nn.BatchNorm1d(50).cuda()
time_bn(bnb, b)
```

`nvprof` on only BatchNorm1d:
```
==10978== NVPROF is profiling process 10978, command: python test/test_batchnorm1d.py
==10978== Warning: Profiling results might be incorrect with current version of nvcc compiler used to compile cuda app. Compile with nvcc compiler 9.0 or later version to get correct profiling results. Ignore this warning if code is already compiled with the recommended nvcc version 
==10978== Profiling application: python test/test_batchnorm1d.py
==10978== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   51.26%  5.5454ms         1  5.5454ms  5.5454ms  5.5454ms  void cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4>(float, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4>, cudnnTensorStruct, float const *, float, cudnnTensorStruct*, float, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const *, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const , double, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4>*, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const *, double, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const *, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const *)
                   42.62%  4.6111ms         7  658.72us     768ns  4.6057ms  [CUDA memcpy HtoD]
                    3.81%  411.87us         1  411.87us  411.87us  411.87us  void cudnn::detail::bn_pointwise<float, float, bool=1, int=2>(float, cudnn::detail::bn_pointwise<float, float, bool=1, int=2>, cudnnTensorStruct, float const *, float, cudnnTensorStruct*, float, cudnn::detail::bn_pointwise<float, float, bool=1, int=2> const *, cudnn::detail::bn_pointwise<float, float, bool=1, int=2> const , double, cudnn::detail::bn_pointwise<float, float, bool=1, int=2>*, cudnn::detail::bn_pointwise<float, float, bool=1, int=2> const *, bool)
                    2.26%  244.12us         1  244.12us  244.12us  244.12us  _ZN2at6native6modern29vectorized_elementwise_kernelILi4EZNS0_23gpu_kernel_with_scalarsIZZZNS0_15add_kernel_cudaERNS_14TensorIteratorEN3c106ScalarEENKUlvE_clEvENKUlvE2_clEvEUlffE_EEvS5_RKT_EUlfE0_NS_6detail5ArrayIPcLi2EEEEEviT0_T1_
                    0.03%  3.5840us         4     896ns     832ns  1.0880us  [CUDA memset]
                    0.02%  2.1120us         1  2.1120us  2.1120us  2.1120us  _ZN2at6native6legacy18elementwise_kernelILi128ELi4EZNS0_15gpu_kernel_implIZNS0_23gpu_kernel_with_scalarsIZZZNS0_15add_kernel_cudaERNS_14TensorIteratorEN3c106ScalarEENKUlvE_clEvENKUlvE4_clEvEUlllE_EEvS6_RKT_EUllE0_EEvS6_SE_EUliE2_EEviT1_
      API calls:   88.94%  2.07116s        11  188.29ms  5.9210us  2.07062s  cudaMalloc
                   10.48%  244.08ms         4  61.020ms     751ns  171.33ms  cudaFree
                    0.25%  5.8220ms         1  5.8220ms  5.8220ms  5.8220ms  cudaDeviceSynchronize
                    0.20%  4.7532ms         6  792.20us  4.7890us  4.7160ms  cudaMemcpyAsync
                    0.04%  1.0101ms         3  336.70us  243.18us  517.24us  cuDeviceTotalMem
                    0.03%  654.78us         1  654.78us  654.78us  654.78us  cudaHostAlloc
                    0.02%  396.75us       285  1.3920us     130ns  67.767us  cuDeviceGetAttribute
                    0.01%  235.19us         2  117.60us  115.91us  119.28us  cudaGetDeviceProperties
                    0.00%  107.83us         4  26.958us  8.0050us  58.229us  cudaLaunchKernel
                    0.00%  96.242us         3  32.080us  18.094us  41.128us  cuDeviceGetName
                    0.00%  86.552us         6  14.425us  4.7490us  61.676us  cudaStreamSynchronize
                    0.00%  83.427us         4  20.856us  1.7040us  77.646us  cudaStreamCreateWithPriority
                    0.00%  55.812us       107     521ns     361ns  2.5750us  cudaGetDevice
                    0.00%  30.107us        44     684ns     420ns  2.7850us  cudaSetDevice
                    0.00%  26.680us        30     889ns     611ns  3.0250us  cudaEventCreateWithFlags
                    0.00%  21.190us         8  2.6480us  1.5730us  9.4880us  cudaStreamCreateWithFlags
                    0.00%  19.057us         4  4.7640us  2.9960us  8.8070us  cudaMemsetAsync
                    0.00%  14.187us         1  14.187us  14.187us  14.187us  cudaMemcpy
                    0.00%  13.749us        29     474ns     351ns  1.1420us  cudaDeviceGetAttribute
                    0.00%  9.5480us         3  3.1820us  2.5450us  3.7870us  cudaEventRecord
                    0.00%  4.3560us        17     256ns     170ns     782ns  cudaGetDeviceCount
                    0.00%  4.3080us         5     861ns     220ns  2.4850us  cuDeviceGetCount
                    0.00%  3.0260us         2  1.5130us     722ns  2.3040us  cudaEventDestroy
                    0.00%  2.8260us         1  2.8260us  2.8260us  2.8260us  cuDeviceGetPCIBusId
                    0.00%  2.3740us         2  1.1870us     851ns  1.5230us  cuInit
                    0.00%  1.9630us         4     490ns     210ns     962ns  cuDeviceGet
                    0.00%  1.9440us         1  1.9440us  1.9440us  1.9440us  cudaHostGetDevicePointer
                    0.00%  1.0210us         4     255ns     190ns     361ns  cudaGetLastError
                    0.00%     962ns         1     962ns     962ns     962ns  cudaDeviceGetStreamPriorityRange
                    0.00%     751ns         2     375ns     371ns     380ns  cuDriverGetVersion
                    0.00%     673ns         3     224ns     181ns     291ns  cuDeviceGetUuid
```

`nvprof` on only BatchNorm2d:
```
==11008== NVPROF is profiling process 11008, command: python test/test_batchnorm1d.py
==11008== Warning: Profiling results might be incorrect with current version of nvcc compiler used to compile cuda app. Compile with nvcc compiler 9.0 or later version to get correct profiling results. Ignore this warning if code is already compiled with the recommended nvcc version
==11008== Profiling application: python test/test_batchnorm1d.py
==11008== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   87.98%  4.6412ms         7  663.04us     768ns  4.6359ms  [CUDA memcpy HtoD]
                    7.30%  384.99us         1  384.99us  384.99us  384.99us  void cudnn::detail::bn_fw_tr_1C11_kernel_NCHW<float, float, int=512, bool=1, int=1>(cudnnTensorStruct, float const *, cudnn::detail::bn_fw_tr_1C11_kernel_NCHW<float, float, int=512, bool=1, int=1>, cudnnTensorStruct*, float const *, float const , cudnnTensorStruct*, cudnnTensorStruct*, cudnnTensorStruct**, float const *, float const *, float const *, cudnnTensorStruct*, cudnnTensorStruct*)
                    4.61%  243.13us         1  243.13us  243.13us  243.13us  _ZN2at6native6modern29vectorized_elementwise_kernelILi4EZNS0_23gpu_kernel_with_scalarsIZZZNS0_15add_kernel_cudaERNS_14TensorIteratorEN3c106ScalarEENKUlvE_clEvENKUlvE2_clEvEUlffE_EEvS5_RKT_EUlfE0_NS_6detail5ArrayIPcLi2EEEEEviT0_T1_
                    0.07%  3.8080us         4     952ns     832ns  1.0880us  [CUDA memset]
                    0.04%  2.3680us         1  2.3680us  2.3680us  2.3680us  _ZN2at6native6legacy18elementwise_kernelILi128ELi4EZNS0_15gpu_kernel_implIZNS0_23gpu_kernel_with_scalarsIZZZNS0_15add_kernel_cudaERNS_14TensorIteratorEN3c106ScalarEENKUlvE_clEvENKUlvE4_clEvEUlllE_EEvS6_RKT_EUllE0_EEvS6_SE_EUliE2_EEviT1_
      API calls:   89.22%  2.06939s        11  188.13ms  6.0610us  2.06885s  cudaMalloc
                   10.42%  241.58ms         4  60.395ms     591ns  166.46ms  cudaFree
                    0.21%  4.7775ms         6  796.26us  5.3700us  4.7402ms  cudaMemcpyAsync
                    0.05%  1.1627ms         3  387.56us  243.20us  670.09us  cuDeviceTotalMem
                    0.03%  649.94us         1  649.94us  649.94us  649.94us  cudaHostAlloc
                    0.03%  609.27us       285  2.1370us     130ns  199.53us  cuDeviceGetAttribute
                    0.02%  380.19us         2  190.09us  144.76us  235.42us  cudaGetDeviceProperties
                    0.01%  251.63us         1  251.63us  251.63us  251.63us  cudaDeviceSynchronize
                    0.00%  111.94us         3  37.313us  23.323us  56.988us  cudaLaunchKernel
                    0.00%  107.57us         3  35.857us  19.146us  45.796us  cuDeviceGetName
                    0.00%  92.584us         4  23.146us  1.6530us  86.974us  cudaStreamCreateWithPriority
                    0.00%  86.432us         6  14.405us  4.8490us  61.626us  cudaStreamSynchronize
                    0.00%  56.987us       107     532ns     350ns  3.7470us  cudaGetDevice
                    0.00%  31.887us        30  1.0620us     591ns  7.2540us  cudaEventCreateWithFlags
                    0.00%  30.469us        44     692ns     420ns  3.0060us  cudaSetDevice
                    0.00%  21.533us         8  2.6910us  1.6230us  9.2780us  cudaStreamCreateWithFlags
                    0.00%  18.986us         4  4.7460us  3.0760us  8.7460us  cudaMemsetAsync
                    0.00%  17.253us         1  17.253us  17.253us  17.253us  cudaMemcpy
                    0.00%  13.552us        29     467ns     351ns  1.4730us  cudaDeviceGetAttribute
                    0.00%  10.539us         3  3.5130us  2.5650us  4.4480us  cudaEventRecord
                    0.00%  4.3450us        17     255ns     140ns  1.2320us  cudaGetDeviceCount
                    0.00%  4.1870us         5     837ns     220ns  2.6950us  cuDeviceGetCount
                    0.00%  2.9260us         2  1.4630us     812ns  2.1140us  cudaEventDestroy
                    0.00%  2.8560us         1  2.8560us  2.8560us  2.8560us  cuDeviceGetPCIBusId
                    0.00%  2.4450us         2  1.2220us     892ns  1.5530us  cuInit
                    0.00%  2.3450us         1  2.3450us  2.3450us  2.3450us  cudaHostGetDevicePointer
                    0.00%  2.0450us         4     511ns     221ns  1.0820us  cuDeviceGet
                    0.00%     932ns         1     932ns     932ns     932ns  cudaDeviceGetStreamPriorityRange
                    0.00%     902ns         3     300ns     190ns     521ns  cuDeviceGetUuid
                    0.00%     721ns         3     240ns     200ns     280ns  cudaGetLastError
                    0.00%     500ns         2     250ns     230ns     270ns  cuDriverGetVersion
```

**Comment by ezyang** [comment]
Relevant lines:

```
 GPU activities:   51.26%  5.5454ms         1  5.5454ms  5.5454ms  5.5454ms  void cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4>(float, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4>, cudnnTensorStruct, float const *, float, cudnnTensorStruct*, float, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const *, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const , double, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4>*, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const *, double, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const *, cudnn::detail::bn_fw_tr_1CHW_kernel_new<float, float, bool=1, int=4> const *)
```

vs 

```
                    7.30%  384.99us         1  384.99us  384.99us  384.99us  void cudnn::detail::bn_fw_tr_1C11_kernel_NCHW<float, float, int=512, bool=1, int=1>(cudnnTensorStruct, float const *, cudnn::detail::bn_fw_tr_1C11_kernel_NCHW<float, float, int=512, bool=1, int=1>, cudnnTensorStruct*, float const *, float const , cudnnTensorStruct*, cudnnTensorStruct*, cudnnTensorStruct**, float const *, float const *, float const *, cudnnTensorStruct*, cudnnTensorStruct*)
```

Which strongly suggests that, at least on CUDA, this is a cuDNN problem.

**Comment by prclibo** [comment]
@ezyang but I guess cuDNN does not explain why the CPU version performs differently?

**Comment by ezyang** [comment]
I'm guessing that there is a crappy 1d batchnorm implementation, but it will be hard to say without a trace.

**Comment by ezyang** [comment]
Unless we can implement Batchnorm1d in terms of Batchnorm2d, this isn't really actionable on our side (for the CUDA difference); maybe our fine friends at NVIDIA can fix this though.

**Comment by vadimkantorov** [comment]
(Maybe this is somehow related: https://github.com/pytorch/pytorch/issues/37713)

**Comment by mruberry** [comment]
Lowering priority. We're not sure how natural this example is.

**Comment by vadimkantorov** [comment]
LxC example is very natural, it's audio and other long-time, few-channels signals

**Comment by rusty1s** [comment]
I and other users of PyTorch Geometric recently noticed the runtime issues with `torch.nn.BatchNorm1d` as well (https://github.com/rusty1s/pytorch_geometric/issues/2394). I personally always wondered why applying batch normalization in Graph Neural Networks make things so slow.

In GNNs, we do not have a "batch" dimension but rather want to normalize across the node dimension within a node feature matrix of shape `[num_nodes, num_features]`. In general, `num_nodes >> num_features`. Manually implementing a `BatchNorm` module for this specific use-case can be up to 15 times faster than using regular `torch.nn.BatchNorm1d`.

Given the popularity of implementing GNNs in PyTorch, I would be very happy to see this addressed :)

**Comment by ngimel** [comment]
Bumping priority based on user activity

**Comment by ngimel** [comment]
reopening, as backward performance is still not fixed
